cmd/internal/database/meilisearch/meilisearch.go

```diff
@@ -0,0 +1,329 @@
+
+	err = retry.Do(func() error {
+		dumpTask, err := db.client.GetTask(dumpResponse.TaskUID)
+		if err != nil {
+			return err
+		}
+		if dumpTask.Status != meilisearch.TaskStatusSucceeded {
+			return fmt.Errorf("dump still processing")
+		}
+
+		db.log.Infow("dump finished", "duration", dumpTask.Duration, "details", dumpTask.Details)
+		return nil
+	}, retry.Attempts(100), retry.Context(ctx))
+	if err != nil {
+		return err
+	}
+	err = db.moveDumpToBackupDir(ctx)
+	if err != nil {
+		return err
+	}
+
+	db.log.Debugw("successfully took backup of meilisearch")
+
+	return nil
+}
+
+// Check implements database.Database.
+func (db *Meilisearch) Check(_ context.Context) (bool, error) {
+	empty, err := utils.IsEmpty(db.dbdir)
+	if err != nil {
+		return false, err
+	}
+	if empty {
+		db.log.Info("data directory is empty")
+		return true, err
+	}
+
+	return false, nil
+}
+
+// Probe implements database.Database.
+func (db *Meilisearch) Probe(_ context.Context) error {
+	_, err := db.client.Version()
+	if err != nil {
+		return fmt.Errorf("connection error:%w", err)
+	}
+	return nil
+}
+
+// Recover implements database.Database.
+func (db *Meilisearch) Recover(ctx context.Context) error {
+	dump := path.Join(constants.RestoreDir, latestStableDump)
+	if _, err := os.Stat(dump); os.IsNotExist(err) {
+		return fmt.Errorf("restore file not present: %s", dump)
+	}
+	start := time.Now()
+
+	if err := utils.RemoveContents(db.dbdir); err != nil {
+		return fmt.Errorf("could not clean database data directory: %w", err)
+	}
+
+	err := db.importDump(ctx, dump)
+	if err != nil {
+		return fmt.Errorf("unable to recover %w", err)
+	}
+
+	db.log.Infow("recovery done", "duration", time.Since(start))
+	return nil
+}
+
+// Upgrade implements database.Database.
+func (db *Meilisearch) Upgrade(ctx context.Context) error {
+	start := time.Now()
+
+	versionFile := path.Join(db.dbdir, meilisearchVersionFile)
+	if _, err := os.Stat(versionFile); errors.Is(err, fs.ErrNotExist) {
+		db.log.Infof("%q is not present, no upgrade required", versionFile)
+		return nil
+	}
+
+	dbVersion, err := db.getDatabaseVersion(versionFile)
+	if err != nil {
+		return err
+	}
+	meilisearchVersion, err := db.getBinaryVersion(ctx)
+	if err != nil {
+		return err
+	}
+	if (dbVersion.Major() == meilisearchVersion.Major()) && (dbVersion.Minor() == meilisearchVersion.Minor()) {
+		db.log.Infow("no version difference, no upgrade required", "database-version", dbVersion, "binary-version", meilisearchVersion)
+		return nil
+	}
+	if dbVersion.GreaterThan(meilisearchVersion) {
+		db.log.Errorw("database is newer than meilisearch binary, aborting", "database-version", dbVersion, "binary-version", meilisearchVersion)
+		return fmt.Errorf("database is newer than meilisearch binary")
+	}
+	if _, err := os.Stat(db.latestStableDumpDst); errors.Is(err, fs.ErrNotExist) {
+		return fmt.Errorf("%q is not present, no upgrade possible, maybe no backup was running before", db.latestStableDumpDst)
```

Comment:
If this happens we are in a loop. Maybe we can add to the error: "please use again previous database image version and wait for a backup, then retry". Or something like this.

However the requirement to have this latest dump is a bit problematic. Imagine a backup schedule of an hour. Somewhere within this period, an operator rolls out a new image version. Then you can potentially loose up to one hour of data.

In the postgres upgrade, the approach is much better because we take the dump in the upgrade process and then restore it. So, I wonder, did you try the same trick with copying the previous binaries already?

Reply:
Fixed the wording, we cannot guard against some race conditions you describe, and loosing one hour is in my opinion good enough.
There is no similar approach available for meilisearch as it is for postgres

Reply:
also mentioned in a TODO in the description